File: sort/merge-sort.py
import random

def merge_sort(arr, d = 0):
    l = len(arr)
    if l == 1:
        return arr
    else:
        left = merge_sort(arr[:l//2], d + 1)
        right = merge_sort(arr[l//2:], d + 1)
        merged = []
        while len(left) > 0 and len(right) > 0:
            src = left if left[0] < right[0] else right
            merged.append(src[0])
            del src[0]
        return merged + left + right

def merge_sort2(arr):
    def merge(a, b, d):
        print(("    " * (d + 1)) + "Merge -> " + str(a) + ", " + str(b))
        r = []
        while len(a) > 0 and len(b) > 0:
            s = a if a[0] < b[0] else b
            r.append(s[0])
            del s[0]
        return r + a + b

    def split(arr, d = -1):
        d += 1
        print(("    " * d) + "Split -> " + str(arr))
        l = len(arr)
        if l == 1:
            return arr
        elif l == 2:
            return merge([arr[0]], [arr[1]], d)
        else:
            left = arr[:l//2]
            right = arr[l//2:]
            return merge(split(left, d), split(right, d), d)

    return split(arr)

print(merge_sort([ random.randint(-99,99) for _ in range(15) ]))

File: sort/merge-sort.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <time.h>

void dump_arr(int *arr, int len)
{
    int i;
    printf("[");
    for (i = 0; i < len; i++)
        printf("%d%s", arr[i], i != len - 1 ? ", " : "]");
}

void merge_sort(int *arr, int *buffer, int len)
{
    if (len > 1) {
        int left_len = len / 2, right_len = len - left_len;
        int *left = arr, *right = arr + left_len;
        int *left_buffer = buffer, *right_buffer = buffer + left_len;

        merge_sort(left, left_buffer, left_len);
        merge_sort(right, right_buffer, right_len);

        int i = 0, l = 0, r = 0;
        while (l < left_len && r < right_len) {
            if (left[l] < right[r]) {
                buffer[i++] = left[l++];
            } else {
                buffer[i++] = right[r++];
            }
        }
        while (l < left_len) {
            buffer[i++] = left[l++];
        }
        while (r < right_len) {
            buffer[i++] = right[r++];
        }
        for (i = 0; i < len; i++) {
            arr[i] = buffer[i];
        }
    }
}

void init_rand_seed()
{
    int seed;
    FILE *fp = fopen("/dev/urandom", "rb");
    fread(&seed, sizeof(seed), 1, fp);
    fclose(fp);
    srand(time(NULL) ^ seed);
}

void test_merge_sort(int size)
{
    int i, *arr, *buffer;
    arr = (int *)malloc(size * sizeof(int));
    buffer = (int *)malloc(size * sizeof(int));
    FILE *fp = fopen("/dev/urandom", "rb");
    fread(arr, sizeof(int), size, fp);
    fclose(fp);
    // for (i = 0; i < size; i++)
    //     arr[i] = rand();
    merge_sort(arr, buffer, size);
    bool go = true;
    for (i = 1; go && i < size; i++)
        if (arr[i - 1] > arr[i])
            go = false;
    free(arr);
    free(buffer);

    printf("Test %d ... %s\n", size, go ? "pass" : "failed");
    fflush(stdout);
}

int main ()
{
    init_rand_seed();
    test_merge_sort(10);
    test_merge_sort(100);
    test_merge_sort(1000);
    test_merge_sort(10000);
    test_merge_sort(100000);
    test_merge_sort(1000000);
    return 0;
}
